drivers/flash/soc_flash_nrf_rram.c

```diff
@@ -0,0 +1,301 @@
+#else
+#define FLASH_SLOT_WRITE 8000 /* longest write takes 7107 us */
+#endif
+
+static int write_op(void *context); /* instance of flash_op_handler_t */
+static int write_synchronously(off_t addr, const void *data, size_t len);
+
+#endif /* !CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE */
+
+static inline bool is_within_bounds(off_t addr, size_t len, off_t boundary_start,
+				    size_t boundary_size)
+{
+	return (addr >= boundary_start && (addr < (boundary_start + boundary_size)) &&
+		(len <= (boundary_start + boundary_size - addr)));
+}
+
+#if WRITE_BUFFER_ENABLE
+static void commit_changes(size_t len)
+{
+	if (nrf_rramc_empty_buffer_check(NRF_RRAMC)) {
+		/* The internal write-buffer has been committed to RRAM and is now empty. */
+		return;
+	}
+
+	if ((len % (WRITE_BUFFER_MAX_SIZE)) == 0) {
+		/* Our last operation was buffer size-aligned, so we're done. */
+		return;
+	}
+
+	nrf_rramc_task_trigger(NRF_RRAMC, NRF_RRAMC_TASK_COMMIT_WRITEBUF);
+
+	barrier_dmem_fence_full();
+}
+#endif
+
+static void rram_write(off_t addr, const void *data, size_t len)
+{
+	nrf_rramc_config_t config = {.mode_write = true, .write_buff_size = WRITE_BUFFER_SIZE};
+
+	nrf_rramc_config_set(NRF_RRAMC, &config);
+
+	if (data) {
+		memcpy((void *)addr, data, len);
+	} else {
+		memset((void *)addr, ERASE_VALUE, len);
+	}
+
+	barrier_dmem_fence_full(); /* Barrier following our last write. */
+
+#if WRITE_BUFFER_ENABLE
+	commit_changes(len);
+#endif
+
+	config.mode_write = false;
+	nrf_rramc_config_set(NRF_RRAMC, &config);
+}
+
+#ifndef CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE
+static void shift_write_context(uint32_t shift, struct flash_context *w_ctx)
+{
+	w_ctx->flash_addr += shift;
+
+	/* NULL data_addr => erase emulation request*/
+	if (w_ctx->data_addr) {
+		w_ctx->data_addr += shift;
+	}
+
+	w_ctx->len -= shift;
+}
+
+static int write_op(void *context)
+{
+	struct flash_context *w_ctx = context;
+	size_t len;
+
+	uint32_t i = 0U;
+
+	if (w_ctx->enable_time_limit) {
+		nrf_flash_sync_get_timestamp_begin();
+	}
+
+	while (w_ctx->len > 0) {
+		len = (WRITE_BUFFER_MAX_SIZE < w_ctx->len) ? WRITE_BUFFER_MAX_SIZE : w_ctx->len;
+
+		rram_write(w_ctx->flash_addr, (const void *)w_ctx->data_addr, len);
+
+		shift_write_context(len, w_ctx);
+
+		if (w_ctx->len > 0) {
+			i++;
+
+			if (w_ctx->enable_time_limit) {
+				if (nrf_flash_sync_check_time_limit(i)) {
+					return FLASH_OP_ONGOING;
+				}
+			}
+		}
+	}
+
+	return FLASH_OP_DONE;
+}
+
+static int write_synchronously(off_t addr, const void *data, size_t len)
+{
+	struct flash_context context = {
+		.data_addr = (uint32_t)data,
+		.flash_addr = addr,
+		.len = len,
+		.enable_time_limit = 1 /* enable time limit */
+	};
+
+	struct flash_op_desc flash_op_desc = {.handler = write_op, .context = &context};
+
+	nrf_flash_sync_set_context(FLASH_SLOT_WRITE);
+	return nrf_flash_sync_exe(&flash_op_desc);
+}
+
+#endif /* !CONFIG_SOC_FLASH_NRF_RADIO_SYNC_NONE */
+
+static int nrf_write(off_t addr, const void *data, size_t len)
+{
```

Comment:
check for `data != NULL` and return -EINVAL, this keeps the meaning of `data == NULL` for delete internal to the driver.